graphics.c

```diff
@@ -0,0 +1,158 @@
+#include "graphics.h"
+#include "display.h"
+
+static inline void vGraphicsPixelWrite(color pixelColor)
+{
+    vDisplayPixelWrite(pixelColor.colorSeperate.redValue, pixelColor.colorSeperate.greenValue, pixelColor.colorSeperate.blueValue);
+}
+
+
+curserStruct curserA,curserB;
+
+void graphicsInit()
+{
+
+vDisplayClear();    //blackout the display
+
+curserA.curserColor=ACURSERCOLOR;
+curserB.curserColor=BCURSERCOLOR;
+
+printGrid();    //print the grid
+
+}
+
+//printRectangle prints a Rectangle
+
+void printRectangle(uint16_t xStart,uint16_t xEnd,uint16_t yStart,uint16_t yEnd,color lineColor)
+{
+    uint16_t row,column;
+
+    vDisplayWindowSet(xStart,xEnd,yStart,yEnd);
+    vDisplayStartPixelWrite();
+
+    for(row=yStart;row<=yEnd;row++)
+    {
+        for(column=xStart;column<=xEnd;column++)
+        {
+            vGraphicsPixelWrite(lineColor);
+        }
+    }
+}
+
+
+//printLineBresenham prints a line between two points, including the second point, excluding the first point
+void printLineBresenham(uint16_t xA,uint16_t yA,uint16_t xB,uint16_t yB,color lineColor)
+{
+uint16_t x=xA,y=yA,dX=xB-xA,dY;
+int16_t delta,step=1;
+uint8_t risingLine=1;
+
+if(yB-yA > 0)   //decision if line shall be printed in negative or positive y-direction
+{
+    dY=yB-yA;
+}
+else
+{
+    dY=yA-yB;
+    risingLine=0;
+    step=-1;
+}
+
+
+delta=dY>>1;
+
+while(y!=yB)
+{
+    delta-=dX;
+
+    y+=step;
+
+    if(delta<=0)
+    {
+        delta+=dY;
+        x++;
+    }
+
+    vDisplayWindowSet(x,x,271-y,271-y);
+    vDisplayStartPixelWrite();
+    vGraphicsPixelWrite(lineColor);
+}
+
+}
+
+/* printWave prints the wave (data saved in printBuffer) from xStart to xEnd. In cases where it is needed it uses the bresenHam algorithm to print lines.
+ * PREVIOUS WAVES NEED TO BE DELETED MANUALY*/
+
+void printWave(uint8_t* printBuffer,uint16_t xStart,uint16_t xEnd,color waveColor)
+{
+
+int16_t dY;
+uint16_t xCounter;
+uint8_t bresenhamEnable;
+
+for(xCounter=xStart;xCounter<=xEnd;xCounter++)
+{
+
+bresenhamEnable=0;
+if(xCounter!=0)
+{
+    dY=printBuffer[xCounter]-printBuffer[xCounter-1];
+    if(dY > 2 || dY < -2)
+    {
+        bresenhamEnable=1;
+    }
+}
+
+if(bresenhamEnable)
+{
+    printLineBresenham(xCounter-1,printBuffer[xCounter-1],xCounter,printBuffer[xCounter],waveColor);
+}
+else
+{
+    vDisplayWindowSet(xCounter,xCounter,271-printBuffer[xCounter],271-printBuffer[xCounter]);
+    vDisplayStartPixelWrite();
+    vGraphicsPixelWrite(waveColor);
+}
+
+}
+
+}
+
+/*blacks out previous wave from xStart to xEnd, refreshs the grid, load new samples in the printBuffer from xStart to xEnd and prints the whole wave */
+void updateWave(uint16_t* sampleBuffer,uint8_t* printBuffer,uint16_t xStart,uint16_t xEnd,color waveColor)
```

Comment:
- In UI-Modul verschieben
- Aufteilen in `vGraphicsUpdateWave` und `vGraphicsUpdateWaveSingle`
- `vGraphicsUpdateWave` löscht den Zeichenbereich und zeichnet komplett neu
- `vGraphicsUpdateWaveSingle` bekommt den Index im Sample-Puffer übergeben, und zeichnet nur den einen Messwert
- Kommentare hinzufügen

Hintergrund: Im Normalfall sind keine "alten" Messwerte angezeigt, wenn eine Wellenform gezeichnet wird. Der Anzeigebereich wird beim Erreichen des rechten Bildschirmrandes immer gelöscht (zum Löschen entweder komplett löschen, oder über eine Liste mit "alten" Messwerten nur die nötigen Pixel neu setzen.

Reply:
Die Funktion soll im Sampler aufgerufen werden wenn 480 Messungen getätigt wurden. Die löscht den alten Graphen, schreibt die neuen samples in den zu zeichnenden Buffer und zeichnet diese mit printwave. Damit ist auch ermöglicht, dass wir zum Beispiel mit jedem Messwert den Graphen updaten, jeder Pixel der alten Welle also sichtbar ist bis ein neuer Messwert für diese x-Koordinate gegeben ist. Könnte man sich offen halten, je nach dem was einem später besser gefällt. Mir gefiel nicht, dass die Welle bei großer timebase gelöscht wird wenn alle 480 Werte voll sind. Besonders da fände ich es besser, wenn wie eben geschildert jede x-Koordinate einzeln aktualisiert wird.

Reply:
Ok. Auch hier die Funktionsweise der Routine im Kommentar kurz beschreiben.
Den "printBuffer" auch am Besten modulintern verwalten, damit man sich außerhalb von Graphics nicht mehr drum zu kümmern braucht.